subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java

```diff
@@ -47,81 +48,69 @@
     private final TaskExecuter delegate;
     private final TaskOutputsGenerationListener taskOutputsGenerationListener;
     private final TaskOutputOriginFactory taskOutputOriginFactory;
+    private final TaskCachingReasonsListener taskCachingReasonsListener;
     private BuildCache cache;
 
-    public SkipCachedTaskExecuter(TaskOutputOriginFactory taskOutputOriginFactory, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskExecuter delegate) {
+    public SkipCachedTaskExecuter(TaskOutputOriginFactory taskOutputOriginFactory, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskCachingReasonsListener taskCachingReasonsListener, TaskExecuter delegate) {
         this.taskOutputOriginFactory = taskOutputOriginFactory;
         this.buildCacheConfiguration = buildCacheConfiguration;
         this.packer = packer;
         this.taskOutputsGenerationListener = taskOutputsGenerationListener;
         this.delegate = delegate;
+        this.taskCachingReasonsListener = taskCachingReasonsListener;
     }
 
     @Override
     public void execute(final TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
         final Timer clock = Timers.startTimer();
 
         final TaskOutputsInternal taskOutputs = task.getOutputs();
-
-        boolean cacheEnabled;
+        boolean cacheable;
         try {
-            cacheEnabled = taskOutputs.isCacheEnabled();
+            TaskCaching taskCaching = taskOutputs.getTaskCaching();
+            cacheable = taskCaching.isCacheable();
+            state.setCacheable(cacheable);
+            taskCachingReasonsListener.taskCacheable(task, taskCaching);
```

Comment:
Since we are passing the whole `task` object anyway, adding the separate `TaskCaching` type seems a bit redundant.